src/chains/filecoin/filecoin/src/blockchain.ts

```diff
@@ -63,56 +68,90 @@ export default class Blockchain extends Emittery.Typed<
 
     this.ready = false;
 
-    // Create genesis tipset
-    const genesisBlock = new BlockHeader({
-      ticket: new Ticket({
-        // Reference implementation https://git.io/Jt31s
-        vrfProof: this.rng.getBuffer(32)
-      }),
-      parents: [
-        // Both lotus and lotus-devnet always have the Filecoin genesis CID
-        // hardcoded here. Reference implementation: https://git.io/Jt3oK
-        new RootCID({
-          "/": "bafyreiaqpwbbyjo4a42saasj36kkrpv4tsherf2e7bvezkert2a7dhonoi"
-        })
-      ]
-    });
-    this.tipsets.push(
-      new Tipset({
-        blocks: [genesisBlock],
-        height: 0
-      })
-    );
-
     // Create the IPFS server
     this.ipfsServer = new IPFSServer(this.options.chain);
 
-    // Fire up the miner if necessary
-    if (this.options.miner.blockTime > 0) {
-      const intervalMine = () => {
-        this.mineTipset();
-      };
+    this.miningTimeout = null;
+
+    // We set these to null since they get initialized in
+    // an async callback below. We could ignore the TS error,
+    // but this is more technically correct (and check for not null later)
+    this.tipsetManager = null;
+    this.blockHeaderManager = null;
 
-      this.miningTimeout = setInterval(
-        intervalMine,
-        this.options.miner.blockTime * 1000
+    this.#database = new Database(options.database);
+    this.#database.once("ready").then(async () => {
+      this.blockHeaderManager = await BlockHeaderManager.initialize(
+        this.#database.blocks!
+      );
+      this.tipsetManager = await TipsetManager.initialize(
+        this.#database.tipsets!,
+        this.blockHeaderManager
       );
 
-      utils.unref(this.miningTimeout);
-    } else {
-      this.miningTimeout = null;
-    }
+      const recordedGenesisTipset = await this.tipsetManager.getTipsetWithBlocks(
+        0
+      );
+      if (recordedGenesisTipset === null) {
+        // Create genesis tipset
+        const genesisBlock = new BlockHeader({
+          ticket: new Ticket({
+            // Reference implementation https://git.io/Jt31s
+            vrfProof: this.rng.getBuffer(32)
+          }),
+          parents: [
+            // Both lotus and lotus-devnet always have the Filecoin genesis CID
+            // hardcoded here. Reference implementation: https://git.io/Jt3oK
+            new RootCID({
+              "/": "bafyreiaqpwbbyjo4a42saasj36kkrpv4tsherf2e7bvezkert2a7dhonoi"
+            })
+          ]
+        });
+
+        const genesisTipset = new Tipset({
+          blocks: [genesisBlock],
+          height: 0
+        });
+
+        this.tipsetManager.earliest = genesisTipset; // initialize earliest
+        await this.tipsetManager.putTipset(genesisTipset); // sets latest
+        await this.#database.db!.put(
+          "latest-tipset",
+          Quantity.from(0).toBuffer()
```

Comment:
This is the equivalent of `Buffer.from([])` (no data), aka `utils.BUFFER_EMPTY`, where `utils` is a named export of `@ganache/utils` (sorry about the sloppy export here :sweat_smile:). If you want `Buffer.from([])` you can use `BUFFER_EMPTY`, and for `Buffer.from([0])` there is a `BUFFER_ZERO`.

Quantity/Data are _intended_ to be used to convert to/from json-rpc QUANTITY/DATA.

Note: these instances aren't frozen, so someone could theoretically do `BUFFER_ZERO[0] = {not zero};`, so these instances should never be used outside ganache internals.

Not requesting a change here, just making sure it is returning what you expect.

Reply:
Since `Buffer.from([...numbers])` expects `numbers` to each individually only be 8 bits, I think using some sort of serializable format to create our buffer array is a better approach. And I want it the initialization to be consistent with the further setting.

FWIW, I did check, but `Quantity.from(0).toBuffer()` is the equivalent of `Buffer.from([0])`:

```
> utils.Quantity.from(0)
0
> utils.Quantity.from(0).toBuffer()
<Buffer 00>
> Buffer.from([])
<Buffer >
> Buffer.from([0])
<Buffer 00>
```

I believe this is exactly what I want :) Thanks for making me double check though!

Reply:
Ah, well.... that's a bug. I guess the fix for that is only in my forking branch right now. Sorry!. It should return `BUFFER_EMPTY`, and it will in `develop` in the future (as ethereum's rlp encoding expects `0 == Buffer.from([])`.

Again, sorry I made you double check for no good reason.

Reply:
We talked about this, and I'm going to change this to use `utils.uintToBuffer`